### ISSUE TITLE
🐛 Biome panics on valid TypeScript constructor overloads

### ISSUE THREAD

**Issue by xaviergonz** [biomejs/biome#9492]
### Environment information

<details>

```bash

  Version:                      2.4.7
  Color support:                false

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     1
  TERM:                         dumb
  JS_RUNTIME_VERSION:           v25.8.1
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Status:                       Not set
  Path:                         unset

Workspace:
  Open Documents:               0
```

</details>

### What happened?

1. Create a file named `constructor-overload.ts` with this content:

```ts
export class Example {
  constructor(value: string)
  constructor(value: string[])
  constructor(value: string | string[]) {
    void value
  }
}
```

2. Run:

```bash
biome lint constructor-overload.ts
```

3. Biome crashes with an internal error:

```text
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue.
When opening the issue, please provide a minimal reproduction, or identify and share the file/code that triggers it. Without a way to reproduce the error, the error can't be fixed:

Source Location: crates/biome_js_semantic/src/semantic_model/scope.rs:150:38
Thread Name: biome::workspace_worker_0
Message: no entry found for key
Stack Trace: Re-run with `RUST_BACKTRACE=1` to capture the stack trace

constructor-overload.ts internalError/panic  INTERNAL  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × processing panicked: no entry found for key
  
  ! This diagnostic was derived from an internal Biome error. Potential bug, please report it if necessary.
```

4. The same file compiles successfully with TypeScript:

```bash
tsc --noEmit --strict --target es2022 --module esnext constructor-overload.ts
```

5. In my real codebase, replacing the constructor overloads with a single tuple-union rest-parameter signature avoids the Biome crash, so this seems specific to the overload declaration form rather than the runtime logic.

### Expected result

It should not throw an internal error. Biome should lint/check the file normally, or at worst report a regular diagnostic if there is a real issue with the code.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by siketyan** [comment]
Likely duplicate of #9483
